### ISSUE TITLE
Figure out how to reconcile with libc++ libcxx.imp

### ISSUE THREAD

**Issue by kimgr** [include-what-you-use/include-what-you-use#1271]
It turns out libc++ ships with a mapping file for itself: https://github.com/llvm/llvm-project/blob/main/libcxx/include/libcxx.imp.

It makes no sense for us to maintain a separate version of this, so it would be nice to rely on this canonical mapping (closer to the source).

A few open questions:

* Can/should we use the libc++ mappings automatically if they happens to exist?
* Should we drop our version?
* Are they equivalent? If not, which is better?
* Should we look for mappings on the include path?

Inspired by this comment from one of the libc++ maintainers: https://github.com/include-what-you-use/include-what-you-use/commit/5540ff09d853605abd328121381a9b720e4269cd#commitcomment-119816408.

**Comment by kimgr** [comment]
@kilroyd @ldionne So I've let this bake for a while, and I'd like to add a proposal:

* Have IWYU search the include path for `*.imp` mapping files and load them implicitly
* We need to soften at least one assertion to make that work: https://github.com/include-what-you-use/include-what-you-use/blob/26b48e7/iwyu_include_picker.cc#L1272 -- otherwise there's a good chance picking up mappings automatically will lead to crashes the user can't get themselves out of
* On that note, I'd like to add a `-nostdinc`-like switch, e.g. `-Xiwyu -no_mapping_search`, to disable this behavior

Once that's in place we should be able to get rid of both the built-in mappings for libcxx and `libcxx.imp` in the IWYU tree.

I do think it might be prudent to do have IWYU provide mappings for libcxx for a while, since upstream `libcxx.imp` was recently released. I have it on both my Linux and Mac systems, but I've probably installed recent LLVM distros on both of them. @ldionne Do you know if libcxx.imp will make it into the XCode SDK (which I assume is what bundles libc++ on the Mac)?

There's [a lot more use of `.imp` files than I expected](https://github.com/search?q=path%3A*.imp&type=code), so there's a good chance that picking them up automatically from the include path will encourage libraries to ship them which should in turn make IWYU's life much easier.

I'm going to start hacking out small changes to replace assertions with warning logs in the coming days, and hopefully have somethink functional for implicit search pretty soon.

Thoughts? Thank you!

**Comment by kilroyd** [comment]
Conceptually I think this is reasonable.

We might want to be careful about picking up _all_ mapping files though. For example a project may create mapping files to handle its thirdparty libraries, which may have their own in later releases. Alternatively a given mapping file may be intended for internal use. Maybe add a field indicating the file is intended to be picked up automatically? `auto_import: true`.

We should consider whether we want to handle the situation where a set of headers has multiple mappings (from multiple projects using the headers). This would be the case as long as there's only partial take up of IWYU mapping files, and requires coordination between projects when they do start using them. This could handled by having a few fields: A library identifier, an integer priority - or a given library identifier only use the mapping file with the highest priority.

Re the github statistics, I feel like 'imp' is used as an extension for multiple purposes. Would it be worth looking for an alternative?

**Comment by kimgr** [comment]
@kilroyd Thanks!

> Re the github statistics, I feel like 'imp' is used as an extension for multiple purposes. Would it be worth looking for an alternative?

Yeah, I'm not too happy with `.imp` either, but it does seem to be gaining traction. Ideally I'd like to define a more fleshed out config format for IWYU (to cover mappings, pragmas, etc.) and call it `.iwyu`, but for now I'd prefer to ride on `.imp` for a while and see how it plays out.

> We might want to be careful about picking up all mapping files though [...] Maybe add a field indicating the file is intended to be picked up automatically? auto_import: true.

Hmm, yeah, that makes a lot of sense, actually. It doesn't silently start working, but on the other hand it doesn't silently break a lot of existing setups.

> We should consider whether we want to handle the situation where a set of headers has multiple mappings (from multiple projects using the headers).

Can you give an example of what you're thinking about here? I'm thinking a sane library project would only describe mappings for their own headers.

**Comment by kilroyd** [comment]
> I'm thinking a sane library project would only describe mappings for their own headers.

Agreed. I'm was thinking about the intermediate state where projects workaround the fact that headers for libraries they use need a mapping file.

> Can you give an example of what you're thinking about here?

The situation I was thinking about is something like the following. Say we have `projecta` that uses `projectb` and `projectc`. Both `projectb` and projectc` use `projectd` in different ways. `projectb` and `projectc` start using IWYU, and create their own mapping files to handle `projectd` headers.

It's conceivable that the mapping files lead to bad outcomes (either incorrect mappings or bad performance). For example, maybe `projectb` uses a few symbols, so just defines direct symbol mappings. `projectc` uses many symbols, but from a single header - it uses an include mapping pointing the whole sub-path to a single header.

`projecta` uses both projects, and IWYU picks up both mapping files. In the example it effectively gets `projectc` usage (everything redirected to one header). Possibly if both include paths are only in a subset of files, some sources still apply `projectb` usage (as the `projectc` mapping isn't consistently picked up).

At some later point `projectd` adds its own mapping file. The projects need to coordinate in some way to keep IWYU happy.

Thinking further, there are a few reasons we can sidestep this:
* for this to be an issue, b and c are leaking d's headers. Ideally they wouldn't, but I feel this isn't unusual.
* b and c's mappings for d shouldn't be found by `projecta` - they should only be in 'internal' include paths

IWYU can say that `imp` files that get installed only cover the project's own headers. To cover thirdparty library usage, use a separate mapping file that doesn't get installed and is not in an include path, that includes the installed mapping file and any extras for thirdparty libraries.

**Comment by kimgr** [comment]
Thanks!

Yeah, my assumption is that it's very unusual for projects to install any `.imp` files currently, but I could be wrong.

On a hunch, I think mappings for a project's own wellbeing will be different from their recommendation for users (e.g. the project itself is free to include its own private headers), so my instinct was that "private" vs. "public" mappings will be almost entirely disjoint.

But I think the idea of a flag in the YAML to mark an `.imp` file as implicitly picked up seems like a good way for libraries to announce their intention for individual file. And the `ref` directive can be used to compose "public" and "private" files in the event that there is overlap.

**Comment by ldionne** [comment]
> Do you know if libcxx.imp will make it into the XCode SDK (which I assume is what bundles libc++ on the Mac)?

Yes, see [this comment](https://github.com/include-what-you-use/include-what-you-use/commit/5540ff09d853605abd328121381a9b720e4269cd#commitcomment-120929531):

> Yes, the first version of libc++ that includes `libcxx.imp` is in Xcode 15.0, which is currently in Beta. It won't update to the Beta version through the normal update process, you have to go to http://developer.apple.com/ to get that version until it is out of Beta.

**Comment by jashank** [comment]
Hello from #1876! Two observations from there which I'll repost here:

+ The link to `libcxx.imp` in the OP is stale, as it is now generated by the LLVM libc++ build process — https://github.com/llvm/llvm-project/blob/e2e2c50bebc03d2e70eb1b33827cf6258496ed2c/libcxx/include/CMakeLists.txt#L1700-L1707

+ An alternative approach to `-I⟨dir⟩` / `-isystem ⟨dir⟩` could be to establish a `site-⟨tool⟩` / `vendor-⟨tool⟩` directory (a la Perl, Python, CMake, ...), into which other IWYU-mapping-providing libraries would install those files.

    For example, a hypothetical implementation might end up with something like:

    ```shell
    hypothetically% ls -1 /usr/share/include-what-you-use/site-maps/
    libcxx.imp
    clang-intrinsics.clang21.imp
    qt5_4.imp
    qt5_11.imp

    # essentially "who installed / owns this file or directory?"
    hypothetically% rpm -qf /usr/share/include-what-you-use/
    iwyu-0.25-1.fc44.x86_64
    hypothetically% rpm -qf /usr/share/include-what-you-use/site-maps/
    iwyu-0.25-1.fc44.x86_64
    hypothetically% rpm -qf /usr/share/include-what-you-use/site-maps/libcxx.imp
    libcxx-devel-21.1.7-1.fc44.x86_64
    ```

    and (in a world where the `.imp` file was certainly YAML, and where a `from` directive existed) one might have a local IWYU mapping file that says:

    ```yaml
    - from: "libcxx.imp"   # whereupon IWYU goes off to find and import `libcxx.imp`.
    - from: "qt5_11.imp"   # likewise, for the Qt 5.11 mappings currently bundled by IWYU.
    - include: [ ... ]     # and then some local mappings for my random project too.
    ```

    This approach gets exciting with multiple libc++ versions installed — e.g., ROCm's embedded LLVM includes its own, which might be preferable for specific use-cases.

    (Alternatively-alternatively, some sort of search path for existing directories in which one might find other `.imp` files.)

    ((Alternatively-alternatively-alternatively, and as a _reductio ad absurdum_: adopt libkpathsea for finding `.imp` files. As someone who does a nontrivial amount of TeX stuff, this is an interesting thought experiment. But probably don't _actually_ do this.))

My gut feeling is that the "we search for your files" tactic vs., the "we require you install files here" tactic differ mainly in intent of ownership (or, more cynically: who can convince enough people of their relative importance). I offer the "we require you install files here" notion as a hopefully-interesting contrast; I necessarily cannot reasonably judge where IWYU falls on that range, and I don't know which of these fits IWYU's model better. (My gut feeling is: IWYU requires buy-in already, so maybe this isn't as high a hurdle to clear; but I might be entirely wrong.)

**Comment by kimgr** [comment]
@jashank Thank you!

The `site-` or `vendor-` extension points look nice, but I'm a bit uneasy about any mechanism where users may have to modify their install root to add extensions. E.g. some people will have a project-local `vendor/` directory containing a clone of a third-party, which could contain `.imp` files. And most people work on more than one project at a time, possibly with differently-versioned dependencies.

So I think my instinct is leaning more towards "we search", but there's some challenges in the discussion above whether `.imp` files should self-identify as "implicitly searchable" or not (`auto_import: true`). Being more explicit using a local manifest to say "search only for these particular files" like you suggested would be another nice way to do that.

Hmm, actually, maybe a base manifest could be built into IWYU, something stupid like:
```
const char* allow_search[] = {
  "libcxx.imp", // we know this one publishes with mappings, and it's a C++ standard component
};
```
Then a switch `-Xiwyu -no_mapping_search` could simply reset/don't use that list.

And if users want to extend the base searchable set, we could offer something like the YAML syntax you suggested.

But I think I like the self-identify mode better. It makes it possible for library developers to mindfully say: "this is an official mapping, and you should always use it", instead of requiring that library _users_ configure a bunch of stuff for IWYU to work correctly.
